projects/bootloader/src/config.c

```diff
@@ -0,0 +1,103 @@
+  // the "safe" page is copied over the corrupted page
+  // If both are corrupted then a critical error is returned
+
+  // persist_init pulls the flash page and puts it into the type BootloaderConfig blob
+  status_ok_or_return(persist_init(&s_config_1_persist, BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE,
+                                   &s_config_1_blob, sizeof(BootloaderConfig), false));
+  status_ok_or_return(persist_ctrl_periodic(&s_config_1_persist, false));
+
+  status_ok_or_return(persist_init(&s_config_2_persist, BOOTLOADER_CONFIG_PAGE_2_FLASH_PAGE,
+                                   &s_config_2_blob, sizeof(BootloaderConfig), false));
+  status_ok_or_return(persist_ctrl_periodic(&s_config_2_persist, false));
+
+  uint32_t config_1_check_crc = prv_compute_crc32(&s_config_1_blob);
+  uint32_t config_2_check_crc = prv_compute_crc32(&s_config_2_blob);
+
+  // checks to see if there are differences in crc to detect corruption
+  bool is_config_1_corrupted = !(s_config_1_blob.crc32 == config_1_check_crc);
+  bool is_config_2_corrupted = !(s_config_2_blob.crc32 == config_2_check_crc);
+
+  // memcpy copies one page to the other ("restoring" the corrupted page)
+  if (is_config_1_corrupted && is_config_2_corrupted) {
+    LOG_CRITICAL("ERROR: Both config pages corrupted \n");
+    return STATUS_CODE_INTERNAL_ERROR;
+  } else if (is_config_2_corrupted) {
+    memcpy(&s_config_2_blob, &s_config_1_blob, sizeof(BootloaderConfig));
+    persist_commit(&s_config_2_persist);
+  } else {
+    memcpy(&s_config_1_blob, &s_config_2_blob, sizeof(BootloaderConfig));
+    persist_commit(&s_config_1_persist);
+  }
+  return STATUS_CODE_OK;
+}
+
+void config_get(BootloaderConfig *config) {
+  // Please persist any changes
+  memcpy(config, &s_config_1_blob, sizeof(BootloaderConfig));
+}
+
+StatusCode config_commit(BootloaderConfig *input_config) {
+  // This function sets the current config page 1 to the input config
+  // To prevent corruption, if page 1 is corrupted during the transfer, then
+  // page 2 is immediately copied over to page 1 and an error is returned
+  // If page 1 is not corrupted then page 1 is copied over to page 2
+
+  input_config->crc32 = prv_compute_crc32(input_config);
+
+  // This memcpy copies over the input config onto the config page 1
+  // This copy is commited into the storage, to change s_config_1_blob
+  // a persist_init will be called
+  memcpy(&s_config_1_blob, input_config, sizeof(BootloaderConfig));
+  persist_commit(&s_config_1_persist);
+
+  // This persist_init updates the s_config_1_blob by pulling from the flash
+  status_ok_or_return(persist_init(&s_config_1_persist, BOOTLOADER_CONFIG_PAGE_1_FLASH_PAGE,
+                                   &s_config_1_blob, sizeof(BootloaderConfig), false));
+
+  // checks to see if blob 1 and input config are different to detect corruption
+  bool config_1_corrupted =
+      !(memcmp(&s_config_1_blob, input_config, sizeof(BootloaderConfig)) == 0);
```

Comment:
nitpick: you can just do `(xyz != 0)` instead of `!(xyz == 0)`, might be clearer